cylc/flow/data_store_mgr.py

```diff
@@ -1002,6 +985,139 @@ def generate_ghost_family(self, fp_id, child_fam=None, child_task=None):
+
+        if itask.tdef.clocktrigger_offset is not None:
+            tproxy.clock_trigger.satisfied = itask.is_waiting_clock_done()
+            tproxy.clock_trigger.time = itask.clock_trigger_time
+            tproxy.clock_trigger.time_string = time2str(
+                itask.clock_trigger_time)
+
+        for trig, satisfied in itask.state.external_triggers.items():
+            ext_trig = tproxy.external_triggers[trig]
+            ext_trig.id = trig
+            ext_trig.satisfied = satisfied
+
+        for label, satisfied in itask.state.xtriggers.items():
+            sig = self.schd.xtrigger_mgr.get_xtrig_ctx(
+                itask, label).get_signature()
+            xtrig = tproxy.xtriggers[sig]
+            xtrig.id = sig
+            xtrig.label = label
+            xtrig.satisfied = satisfied
+            self.xtrigger_tasks.setdefault(sig, set()).add(tproxy.id)
+
+        if tproxy.state in self.latest_state_tasks:
+            tp_ref = Tokens(tproxy.id).relative_id
+            tp_queue = self.latest_state_tasks[tproxy.state]
+            if tp_ref in tp_queue:
+                tp_queue.remove(tp_ref)
+            self.latest_state_tasks[tproxy.state].appendleft(tp_ref)
+
+    def apply_task_proxy_db_history(self):
+        """Extract and apply DB history on given task proxies."""
+
+        if not self.db_load_task_proxies:
+            return
+
+        flow_db = self.schd.workflow_db_mgr.pri_dao
+
+        task_ids = set(self.db_load_task_proxies.keys())
+
+        # Batch load rows with matching cycle & name column pairs.
+        prereq_ids = set()
+        for (
+                cycle, name, flow_nums_str, status, submit_num, outputs_str
+        ) in flow_db.select_tasks_for_datastore(task_ids):
+            tokens = self.id_.duplicate(
+                cycle=cycle,
+                task=name,
+            )
+            itask, is_parent = self.db_load_task_proxies[tokens.relative_id]
+            itask.submit_num = submit_num
+            flow_nums = set(json.loads(flow_nums_str))
+            # Do not set states and outputs for future tasks in flow.
+            if (
+                    itask.flow_nums and
+                    flow_nums != itask.flow_nums and
+                    not is_parent
+            ):
+                itask.state_reset(TASK_STATUS_WAITING)
+                continue
+            else:
+                itask.flow_nums = flow_nums
+                itask.state_reset(status)
+            if (
+                    outputs_str is not None
+                    and itask.state(
+                        TASK_STATUS_RUNNING,
+                        TASK_STATUS_FAILED,
+                        TASK_STATUS_SUCCEEDED
+                    )
+            ):
+                for message in json.loads(outputs_str).values():
+                    itask.state.outputs.set_completion(message, True)
+            # Gather tasks with flow id.
+            prereq_ids.add(f'{tokens.relative_id}/{flow_nums_str}')
```

Comment:
Because `flow_nums` are defined as a set on the TaskProxy there is no guarantee of order when they are written to the DB which could mess with this comparison logic.

We should push serialisation/deserialisation through a sort to make it safe. Seems to work now so we can do this later - https://github.com/cylc/cylc-flow/issues/4672